KEP: TAS - Node replacement and workload eviction for tainted nodes (#8929)
kubernetes-sigs/kueue: keps/2724-topology-aware-scheduling/README.md

evicted by e.g. `waitForPodsReady.recoveryTimeout`). One can limit the number of retries
to only one, by setting the `TASFailedNodeReplacementFailFast` feature gate to `true`.

#### Tainted nodes treatment

##### Since v0.16.1

When the `TASTaintEviction` feature gate is enabled, Kueue treats tainted nodes as unhealthy. This applies to nodes with `NoExecute` taint, or nodes with `NoSchedule` taint where all pods of the workload running on that node are failing or terminating.

- **NoExecute**: Nodes with `NoExecute` taint that is not tolerated by the workload are considered unhealthy. The pods on such nodes are expected to be terminated by the node controller. Once terminated, Kueue will attempt to replace the node if `TASFailedNodeReplacement` is enabled, and evict the workload if no replacement is possible. If `tolerationSeconds` is specified, Kueue waits for the duration before treating the node as unhealthy.

> - **NoSchedule**: Nodes with `NoSchedule` taint that is not tolerated by the workload are considered unhealthy only if the pods of the workload running on them are in a failed state or if they are in pending state. In this case, Kueue triggers node replacement. If the pods are running and healthy, the `NoSchedule` taint does not trigger node replacement, allowing the workload to continue running.

Reviewer: "only if the pods of the workload running on them are in a failed state or if they are in pending state" -- all the pods or some pods?

Author: description updated

evicted by e.g. `waitForPodsReady.recoveryTimeout`). One can limit the number of retries
to only one, by setting the `TASFailedNodeReplacementFailFast` feature gate to `true`.

#### Tainted nodes treatment

##### Since v0.16.1

When the `TASTaintEviction` feature gate is enabled, Kueue treats tainted nodes as unhealthy. This applies to nodes with `NoExecute` taint, or nodes with `NoSchedule` taint where all pods of the workload running on that node are failing, terminating, or in a pending state.

- **NoExecute**: Nodes with `NoExecute` taint that is not tolerated by the workload are considered unhealthy. The pods on such nodes are expected to be terminated by the node controller. Once terminated, Kueue will attempt to replace the node if `TASFailedNodeReplacement` is enabled, and evict the workload if no replacement is possible. If `tolerationSeconds` is specified, Kueue waits for the duration before treating the node as unhealthy.

> - **NoSchedule**: Nodes with `NoSchedule` taint that is not tolerated by the workload are considered unhealthy only if all pods of the workload that have topology assignment to that node are in a failed state or if they are in a pending state. In this case, Kueue can trigger node replacement.

Reviewer 2: ```suggestion
- **NoSchedule**: Nodes with the `NoSchedule` taint, that is not tolerated by the workload, are considered unhealthy only if all pods of the workload that have topology assignment to that node are terminating, in the failed state, or if they are unscheduled. In this case, Kueue can trigger node replacement.
```

evicted by e.g. `waitForPodsReady.recoveryTimeout`). One can limit the number of retries
to only one, by setting the `TASFailedNodeReplacementFailFast` feature gate to `true`.

#### Tainted nodes treatment

We introduce the `TASReplaceNodeOnTaints` feature gate from v0.17 as Beta, and backport to v0.15 and v0.16 as Alpha.
When enabled, Kueue treats tainted nodes as unhealthy. This applies to nodes with `NoExecute` taint,
or nodes with `NoSchedule` taint where all pods of the workload running on that node are failing, terminating, or in unscheduled state.

- **NoExecute**: Nodes with the `NoExecute` taint, that is not tolerated by the workload, are considered unhealthy.
The pods on such nodes are expected to be terminated by the node controller. Once terminated, Kueue will attempt
to replace the node if `TASFailedNodeReplacement` is enabled, and evict the workload if no replacement is possible.
If `tolerationSeconds` is specified, Kueue waits for the duration before treating the node as unhealthy.

> - **NoSchedule**: Nodes with the `NoSchedule` taint, that is not tolerated by the workload, are considered unhealthy

Reviewer 3: ```suggestion
- **NoSchedule**: Nodes with the `NoSchedule` effect taint, that is not tolerated by the workload, are considered unhealthy
```

only if all pods of the workload that have topology assignment to that node are terminating, in the failed state,
or if they are unscheduled. In this case, Kueue can trigger node replacement.
  
For workloads for which a single Node replacement is possible, and the pods bound to the node are unscheduled (no `spec.nodeName` set),
because they cannot run due to a taint, Kueue marks the pods as `Failed` and adds the following condition to the pods:
  ```yaml
  type: TerminatedByKueue
  status: True
  reason: UnschedulableDueToUntoleratedNoScheduleTaint

>   ```

Reviewer 3: As another observability, could we issue an event as well?
The failed pods could be deleted immediately based on high-level workload (e.g., Job) specifications.
So, I think that event is important in this case.

to only one, by setting the `TASFailedNodeReplacementFailFast` feature gate to `true`.

#### Tainted nodes treatment

We introduce the `TASReplaceNodeOnNodeTaints` feature gate from v0.17 as Beta, and backport to v0.15 and v0.16 as Alpha.
When enabled, Kueue treats tainted nodes as unhealthy. This applies to nodes with `NoExecute` taint,
or nodes with `NoSchedule` taint where all pods of the workload running on that node are failing, terminating, or in unscheduled state.

- **NoExecute**: Nodes with the `NoExecute` effect taint, that is not tolerated by the workload, are considered unhealthy.
The pods on such nodes are expected to be terminated by the node controller. Once terminated, Kueue will attempt
to replace the node if `TASFailedNodeReplacement` is enabled, and evict the workload if no replacement is possible.
If `tolerationSeconds` is specified, Kueue waits for the duration before treating the node as unhealthy.
- **NoSchedule**: Nodes with the `NoSchedule` effect taint, that is not tolerated by the workload, are considered unhealthy
only if all pods of the workload that have topology assignment to that node are terminating, in the failed state,

> or if they are unscheduled. In this case, Kueue can trigger node replacement.

Reviewer 3: Thank you for describing that.

1. That sounds reasonable to me. Could you add what "TAS" will do to the NoSchedule section? I believe that the description would be worth it because there are different actions between NoExecute and NoSchedule.

2. SGTM, that was my expectation.

> or if they are unscheduled. In this case, Kueue can trigger node replacement.

Reviewer 3: ```suggestion
or if they are unscheduled. In this case, Kueue can trigger node replacement since the Pods with unhealthy Nodes could be refused due to kube-scheduler Pod scheduling based on Node taints tolerations.
```

@j-skiba what about this extended description?

Author: I added this below
```
If a workload is assigned to a node that is tainted with `NoSchedule`, the pods will remain in a pending state because
  the Kubernetes scheduler will not schedule them on that node.
```
would you like it to be here instead?
